hal/src/thumbv7em/pukcc.rs

```diff
@@ -0,0 +1,579 @@
+#![deny(missing_docs)]
+//! Public Key Cryptography Controller (PUKCC)
+//!
+//! This module provides both low and high level abstractions for dealing with
+//! a PUKCC peripheral.
+//!
+//! PUKCC consists of a set of functions (called services) hidden within a
+//! reserved region of memory. These functions usually make use of a separate
+//! piece of RAM to operate called CryptoRAM.
+//!
+//! [`c_abi`] module contains raw structs and callable C-like function
+//! definitions. [`Pukcc`] wraps this low-level access API and exposes it in a
+//! safe manner.
+
+pub mod c_abi;
+pub mod curves;
+
+use crate::pac::MCLK;
+use c_abi::{u4, CryptoRamSlice, Service};
+use curves::Curve;
+
+/// This macro linearly copies provided iterable slices/arrays to CryptoRAM and
+/// assigns slices to provided declared local variables from outer scope
+///
+/// All parameters are stored in big endian format. PUKCC algorithms require
+/// little endian format. This is solved by copying data in reverse within a
+/// macro.
+macro_rules! copy_to_cryptoram {
+        (
+            $crypto_ram:expr,
+            $(
+                ($name:ident, $data:expr)
+            ),+
+        ) =>
+        {
+            {
+            (&[])
+                .iter()
+            $(
+                .chain($data.iter().rev()) // > Little endian format! <
+            )+
+            .zip($crypto_ram.iter_mut())
+            .for_each(|(data_iter, cr_iter)| *cr_iter = *data_iter);
+
+            let mut _offset = 0;
+            $(
+                $name = &$crypto_ram[_offset.._offset + $data.len()];
+                _offset += $data.len();
+            )+
+            }
+        }
+}
+
+/// Struct representing a PUKCC peripheral.
+///
+/// It provides an access to cryptograhic algorithms in a safe, high-level
+/// manner
+pub struct Pukcc {
+    __: (),
+}
+
+impl Pukcc {
+    /// Constructor.
+    ///
+    /// Waits for a CryptoRAM readiness, enables a synchronous PUKCC clock and
+    /// performs a self test. In case a self test fails it returns an error
+    pub fn enable(mclk: &mut MCLK) -> Result<Self, SelfTestFailure> {
+        // TODO: Alleged RNG initialization. Passthrough constructor?
```

Comment:
How do you mean? Is the RNG initialized? is there a way to check?

The module should fail to operate if it cannot verify a RNG is operating, either because it initialized it itself, or it checked some status registers.

Another way to do it is require some struct be passed in that can only be obtained by initializing the RNG elsewhere.

Reply:
Documentation puts a lot of stress on "having an RNG module enabled compulsorily". Like for example on a page 1382. I don't know if these hidden functions discreetly use RNG related peripherals and this is why you are expected to make sure they are enabled or not. 

ATM I think they are just meant to vaguely remind a programmer to use RNG when necessary while generating input parameters for these services. Like `k` for instance. So probably I should remove this TODO.